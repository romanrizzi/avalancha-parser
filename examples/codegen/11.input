
fun add
  Zero   , y -> y
  Suc(x) , y -> Suc(add(x, y))

fun mul2  x -> add(x, x)

fun binary_to_unary
  : x -> y
  ? is_binary(x)
  ! is_unary(y)
  Z              -> Zero
  B0(x)          -> mul2(binary_to_unary(x))
  B1(x)          -> Suc(mul2(binary_to_unary(x)))

fun is_binary
  Z     -> True
  B0(x) -> is_binary(x)
  B1(x) -> is_binary(x)
  _     -> False

fun is_unary
  Zero   -> True
  Suc(x) -> is_unary(x)
  _      -> False

print binary_to_unary(B0(B0(B1(Z))))

