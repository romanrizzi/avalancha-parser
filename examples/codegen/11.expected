Suc(Suc(Suc(Suc(Zero))))
